### PR TITLE
Forward function update

### PR DESCRIPTION
**Pull request by vcuspinera**
Hi guys, send you my function, but I still have some part to do. I will update it for the Milestone 3

### CODE REVIEW

**Comment by techrah** [comment]
@vcuspinera Please run `devtools::check()` (Build/Check Package) and fix errors and warnings.

File: R/forward_selection.R [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  if (!any(class(X) == "data.frame")) {
+    stop("Expected a 'data.frame' object for `data`.")
+  }
+
+  if (sum(!(dim(X))==1) != 2){
+    stop("X must be a 2-d array")
+  }
+
+  if (!any(class(y) == "data.frame")) {
+    stop("Expected a 'data.frame' object for `data`.")
```
Comment (jackyho112) [inline R/forward_selection.R:47]: Should probably distinguish that from the warning about X. Also, can't y just be a vector?

File: R/forward_selection.R [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  # max_features should be greather or equal to min_features
+  if (min_features > max_features){
+    stop("max_features should be greater or equal to min_features.")
+  }
+
+  # min_features should be positive, we want to select at least one feature
+  if (min_features < 1){
+    stop("min_features should be a positive number.")
```
Comment (jackyho112) [inline R/forward_selection.R]:
```suggestion
    stop("min_features should be more than zero.")
```

File: R/forward_selection.R [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  # Initialize parameters
+  # Obtain initial array of randomly selected features
+  ftr_no_selection <- c(1:length(X))
+  ftr_selection <- c()
+  ftr_running <- c()
+  scores_fn <- c()#Inf
```
Comment (jackyho112) [inline R/forward_selection.R]: What does the comment here mean?

File: R/forward_selection.R [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  X_new <- c()
+  best_scores_iter <- Inf
+  best_scores_all <- c()
+  flag_keep_running = TRUE
+  flag_stop_running = FALSE
+
+  # The algorithm
```
Comment (jackyho112) [inline R/forward_selection.R]: No need for this comment.

File: R/forward_selection.R [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  repeat{ #for (j in 1:max_features){
```
Comment (jackyho112) [inline R/forward_selection.R]: What's this comment?

File: R/forward_selection.R [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      ftr_running[i] <- i
+      scores_fn[i] <- scorer(cbind(X_new, y))
+    }
+
+    # selects the feature with the best performance of each itteration
+    best_scores_iter <- min(scores_fn)
+    selected_ftr <- match(best_scores_iter, scores_fn)
+
+    # flag to stop the algorithm if the score doesn't decrease at least by 5%
+    if (length(ftr_selection) >= 1){
```
Comment (jackyho112) [inline R/forward_selection.R:96]: Combine the two conditions and use `&`

File: R/forward_selection.R [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  # Initialize parameters
+  # Obtain initial array of randomly selected features
+  ftr_no_selection <- c(1:length(X))
+  ftr_selection <- c()
+  ftr_running <- c()
+  scores_fn <- c()#Inf
+  X_new <- c()
+  best_scores_iter <- Inf
+  best_scores_all <- c()
+  flag_keep_running = TRUE
```
Comment (jackyho112) [inline R/forward_selection.R:78]: Why do you need two flags? It is quite confusing. Doesn't `flag_keep_running = TRUE` mean `flag_stop_running = FALSE` inherently? When is it not the case?

File: R/forward_selection.R [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    # break if the the algorithm got more than min_features and
+    # additional features doesn't improve the result
+    if (flag_keep_running == FALSE & flag_stop_running == TRUE){
+      break
+    }
+
+    # break if in reaches the max_features
+    if (length(ftr_selection) >= max_features){
```
Comment (jackyho112) [inline R/forward_selection.R:114]: Redundancy because you have checked that condition right above?

File: R/forward_selection.R (hunk 5 above)
Comment (jackyho112) [reply R/forward_selection.R]: Also, a `while` loop wouldn't work here? It's hard to read when the `break` condition is hidden in the loop block.

File: tests/testthat/test-forward_select.R [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  data <- dplyr::select(tgp::friedman.1.data(), -Ytrue)
+  X <- data[1:(length(data)-1)]
+  y <- data[length(data)]
+
+  # Run Test compare scores with expected ones
+  set.seed(1230)
+  results <- forward_select(scorer, X, y, 4, 4)
+  testthat::expect_setequal(results, c(1, 2, 4, 5))
+
+  # X and y are Dataframes
```
Comment (jackyho112) [inline tests/testthat/test-forward_select.R:38]: A lot of the test cases below should be separate and outside of the relevant feature test.

File: tests/testthat/test-forward_select.R [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+# Test output arrays are not empty
+testthat::test_that("features returned is not empty", {
+  # Create dataframe and remove Ytrue column; keep Y.
+  data <- dplyr::select(tgp::friedman.1.data(), -Ytrue)
+  X <- data[1:(length(data)-1)]
+  y <- data[length(data)]
+
+  # Run Test
+  testthat::expect_gt(length(simulated_annealing(scorer, X, y)), 0)
```
Comment (jackyho112) [inline tests/testthat/test-forward_select.R:79]: What's this doing here?

**Review by jackyho112 — CHANGES_REQUESTED** [review]
A bunch of concerns. Again, don't need to fix all of them. Do what you can. If some are too much to change, let me know and it's ok.

File: R/forward_selection.R (hunk 1 above)
Comment (vcuspinera) [reply R/forward_selection.R:47]: I separate the information of 'X' from 'y', and in this case the algorithm needs data frames for both `X` and `y`.

File: R/forward_selection.R (hunk 3 above)
Comment (vcuspinera) [reply R/forward_selection.R]: infinite; I used it for a previous version of the function but doesn't make sense anymore, so I just erase it.

**Review by jackyho112 — APPROVED** [review]
Looks good!

**Pull request merged**